### PR TITLE
Single duty instance - Fix SSV Tests

### PR DESCRIPTION
**Pull request by GalRogozinski**

### CODE REVIEW

File: ssv/spectest/tests/runner/consensus/future_decided.go [hunk 1]
```diff
@@ -2,16 +2,20 @@ package consensus
...
 	"github.com/herumi/bls-eth-go-binary/bls"
 
 	"github.com/bloxapp/ssv-spec/ssv/spectest/tests"
 	"github.com/bloxapp/ssv-spec/types"
 	"github.com/bloxapp/ssv-spec/types/testingutils"
 )
 
-// FutureDecided tests a running instance at FirstHeight, then processing a decided msg from height 2 and returning decided but doesn't move to post consensus as it's not the same instance decided
+// TODO: should a future decided message stop previous instances? currently we return an error but should we?
+// A future decided means that no one will process our old instance messages...
```
Comment (GalRogozinski) [inline ssv/spectest/tests/runner/consensus/future_decided.go]: @alonmuroch
should open an issue?

File: ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/bloxapp/ssv-spec/types/testingutils"
+)
+
+// DuplicateDutyFinished is a test that runs the following scenario:
+// - Runner is assigned a duty
+// - Runner finishes the duty
+// - Runner is assigned the same duty again
+// TODO - Does it make sense that the runner starts? Does it make sense that duties with precon phase don't return an error?
+func DuplicateDutyFinished() *MultiStartNewRunnerDutySpecTest {
```
Comment (GalRogozinski) [inline ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go]: This Test was added since the original Finished test seems to be trying to cover more than one scenario.

What is important here that the runner finishes and then a duty of a similar height or less appears. It doesn't have to be duplicate so maybe we can change the test.

File: ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/bloxapp/ssv-spec/ssv"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/bloxapp/ssv-spec/types/testingutils"
+)
+
+// DuplicateDutyFinished is a test that runs the following scenario:
+// - Runner is assigned a duty
+// - Runner finishes the duty
+// - Runner is assigned the same duty again
+// TODO - Does it make sense that the runner starts? Does it make sense that duties with precon phase don't return an error?
```
Comment (GalRogozinski) [inline ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go]: @alonmuroch

should I open an issue? it just seems to me that it is odd that duties that can't enter consensus phase return an error so late

File: ssv/spectest/tests/runner/duties/newduty/post_wrong_decided.go [hunk 4]
```diff
@@ -11,9 +11,12 @@ import (
 )
 
 // PostWrongDecided tests starting a new duty after prev was decided wrongly (future decided)
+// This can happen if we receive a future decided message from the network.
 func PostWrongDecided() tests.SpecTest {
 	ks := testingutils.Testing4SharesSet()
 
+	// TODO: not sure what is the value of doing this. We initialize the runner with an impossible decided value.
+	// Maybe we should ensure that `ValidateDecided()` doesn't let the runner enter this state and delete the test?
```
Comment (GalRogozinski) [inline ssv/spectest/tests/runner/duties/newduty/post_wrong_decided.go:19]: @alonmuroch should I open an issue?

File: ssv/spectest/tests/runner/duties/newduty/post_invalid_decided.go [hunk 5]
```diff
@@ -33,15 +33,17 @@ func PostInvalidDecided() tests.SpecTest {
 		return byts
 	}
 
+	// TODO: not sure what is the value of doing this. We initialize the runner with an impossible decided value.
+	// Maybe we should ensure that `ValidateDecided()` doesn't let the runner enter this state and delete the test?
```
Comment (GalRogozinski) [inline ssv/spectest/tests/runner/duties/newduty/post_invalid_decided.go:37]: @alonmuroch should I open an issue?

File: types/testingutils/runner.go [hunk 6]
```diff
@@ -184,6 +184,98 @@ var decideRunner = func(consensusInput *types.ConsensusData, height qbft.Height,
 	return v.DutyRunners[types.BNRoleAttester]
 }
 
+// //////////////////////////////// For SSV Tests ////////////////////////////////////////////////////////////////
+var SSVDecidingMsgsForHeight = func(consensusData *types.ConsensusData, msgIdentifier []byte, height qbft.Height, keySet *TestKeySet) []*qbft.SignedMessage {
```
Comment (alonmuroch) [inline types/testingutils/runner.go:188]: what that copied from somewhere else?

File: ssv/spectest/generate/main.go [hunk 7]
```diff
@@ -72,12 +75,23 @@ func clearStateComparisonFolder() {
 }
 
 func writeJsonStateComparison(name, testType string, post interface{}) {
+	postMap, ok := post.(map[string]types.Root)
+
+	if ok {
+		for subTestName, postState := range postMap {
+			writeSingleSCJson(filepath.Join(name, subTestName), testType, postState)
```
Comment (alonmuroch) [inline ssv/spectest/generate/main.go]: was that done for readability? or?

File: ssv/spectest/tests/runner/consensus/future_decided.go (hunk 1 above)
Comment (alonmuroch) [reply ssv/spectest/tests/runner/consensus/future_decided.go]: I don't think it should, the instance will eventually stop processing on its on

File: ssv/spectest/generate/state_comparison/*synccommitteeaggregator.SyncCommitteeAggregatorProofSpecTest/sync committee aggregator all are aggregators.json [hunk 8]
```diff
@@ -0,0 +1,171 @@
+{
```
Comment (alonmuroch) [inline ssv/spectest/generate/state_comparison/*synccommitteeaggregator.SyncCommitteeAggregatorProofSpecTest/sync committee aggregator all are aggregators.json]: I think my PR shouldn't be adding new state comparison jsons but just change existing ones (so we see what's changed)

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json [hunk 9]
```diff
@@ -3,20 +3,20 @@
 				"State": {
 						"PreConsensusContainer": {
 								"Signatures": {
-										"e48ae6ae7037605000abcd09bfcfd6e20c9d4451a2300c446bfb82debad21fbd": {
-												"1": "qmLVsaiUvU6bjLbsSoPLw4CTNR9leefbu7bLw4j1rYbxEM2ZnnM/iORstNrGYenoCkrSMxRJOZUBBRuZKopbWL++g717SyADz3M3F9cMjLQP8v1zvMCC0SBx/jafPpm0",
-												"2": "mYa/q5etbMWdnpvFoS2SSPeuwEdgBWK7UVcffDmda+NCHKm3/cmZj14Pi7doq6rUEi1UddIg0G9OMtDtOikEI0gcKtzl4yyxa1BY12YTO51kGATVsCI6fH1aPnAox/4U",
-												"3": "iX12XMJaUp6vo3kfrzjXElOivuLT/R2VN4uYN7Uv6PY92Lgx/qc9tXcFJqVjy2SLE8pImhwAUCUsdOjpkzP31KHxQeVzKaIQDECtbM8rXJYdS74JkVoSf9o2CvWeuNtr"
+										"35f2cc4ab8c7b2db07738165f6cbc9d889eb11e6d14fc2e363bd90c100b5fcc2": {
```
Comment (alonmuroch) [inline ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: why did the sigs change?

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json [hunk 10]
```diff
@@ -79,12 +79,12 @@
 										},
 										"ID": "AAADAY6ABmVRqBsxglhwntr33R9jzWhqDk24spu7es/mVghnevWlJ9lEjuR4NUheArULwAIAAAA=",
 										"Round": 1,
-										"Height": 0,
+										"Height": 5193728,
```
Comment (alonmuroch) [inline ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: why such an extreme height?

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json [hunk 11]
```diff
@@ -79,12 +79,12 @@
...
.../ueBx8BakxEj+RAFTQlsLhVJUNMqJsOO23DdpQokK+TRXOYMJldnsUEBHNhKpYXDr3mPwetepj6T4KQmw+FGj0AvfmTyAoGkvXPMEjQXTxdiqaQZifVwmXA2yIWx/b+cgVOh0Zr7ZSahI8pqL+bpjGAJ+EOfa27qiBRTzFjvNEM4/9BHg86cKMLjc4EqZRV2Q/Z57ZmrNfQCTm7jGHflNrA8OGFvzMmTNlFD3a/845yAWzkWdAcOmTxkZBVgQ6hCZoYMdp3iGLtfVpj0x6m3QUJTXLzAjVs/dHy/an37tsewzMWK9IhrxEFVhJbpyE2AZgEXd38BzbhMCi07Dy2Kbqz/Wg5VvRvmOHw3k66PUjHFlpeukUiUpJs64To6Eky/bOozt+tXXPbNE+YHPd9NAzo/hzZsJwia/ANwfvmkTIKDiHM9cPCbwI3ldNB88ZPwomoIzoJTv9sm1TBmMgEsba4ZR4PesxtyrDW71X8H8WZ3RthfXVshMriF+MsCBpSXgPlAYwc5bcfs55OCJfxaVcZe02CMJ/ZoupsIh1l54kllXRWkw9d7FDOrJvVtgOLKkXjYhQHuaCoU0HsXY16ZaNotmKqetKWuY55C7tDHc1MT0zCP0HlYnVyy3Tgakpj2cyWXam3NPjj4NuS8MQaIX2NIxXyQvq3Xy8oXVwuQEwqq3H6w7Sv7l8XIp5HQtrJzbUOw9ESQlEnm6jK6cG3f4oxAeoXoLj8iBkaW0v17N7M+AeABatkQR/lfDPXF1qvI/XaYxBVcKQoaHbhC535lbGmti8Bs6p4A2tvsiGymSvX+MEXryMVJvyPnHGNPAtqLJQYYwWmlTkr0XXmfLeZ0e94BOVoT8aYF6b5ZztBKUed+0HMPNCBXEWTqUkemcLli6/a0U2YHSMpQIAAAAAAAAAAwAAAAAAAAAAAQIDBAUGBwgJCgsMDQ4PEBESEwAAZKeztuANAgAAAAAAAAC4m+vGmXaXJqMYyOmXG9MXEpfGGupKZXinpPlLVH3LpbrBaokQi2tqH+NpXRqHSgsAAQIDBAUGBwgJCgsMDQ4PEBESEwABAgMEBQYHCAkKCwwNDg8QERITFBUWFxgZGhscHR4fICEiIyQlJicoKSorLC0uLzAxMjM0NTY3ODk6Ozw9Pj9AQUJDREVGR0hJSktMTU5PUFFSU1RVVldYWVpbXF1eXw==",
```
Comment (alonmuroch) [inline ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: why did the decided value changed?

File: ssv/spectest/tests/runner/consensus/invalid_decided_value.go [hunk 12]
```diff
@@ -39,19 +40,32 @@ func InvalidDecidedValue() tests.SpecTest {
...
 							[]types.OperatorID{1, 2, 3},
+							qbft.Height(testingutils.TestingDutySlot),
 							testingutils.SyncCommitteeContributionMsgID,
 							consensusDataByts(types.BNRoleSyncCommitteeContribution),
-						), nil),
+						), nil,
+					),
 				},
-				PostDutyRunnerStateRoot: "0e966c2d067b51f57254f5e56acd98673f026c89f3d4df74bd567f9668417501",
+				PostDutyRunnerStateRoot: "aff4af0dbbead81d6cb9dd4ff734d4660712a8b5ab8e9016a3f86e2c2ead7549",
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/consensus/invalid_decided_value.go:68]: if we don't change it will fail? it takes the post root from the json no?

File: ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go [hunk 13]
```diff
@@ -20,7 +20,7 @@ func ValidDecided10Operators() tests.SpecTest {
 				Runner:                  testingutils.SyncCommitteeContributionRunner(ks),
 				Duty:                    &testingutils.TestingSyncCommitteeContributionDuty,
 				Messages:                testingutils.SSVDecidingMsgsV(testingutils.TestSyncCommitteeContributionConsensusData, ks, types.BNRoleSyncCommitteeContribution),
-				PostDutyRunnerStateRoot: "ec47a2194bab81891bd581eaf427af211b53e13d1f77a6703eeb56a693caa878",
+				PostDutyRunnerStateRoot: "a991b6470a8c7a55f4ce89aea91925c2d80a9a8c4258545cc2fb17cabc388719",
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go:23]: why is this changed? don't we take post root from json?

File: ssv/spectest/tests/runner/duties/newduty/consensus_not_started.go [hunk 14]
```diff
@@ -24,45 +24,43 @@ func ConsensusNotStarted() tests.SpecTest {
 		Name: "new duty consensus not started",
 		Tests: []*StartNewRunnerDutySpecTest{
 			{
-				Name:                    "sync committee aggregator",
-				Runner:                  startRunner(testingutils.SyncCommitteeContributionRunner(ks), &testingutils.TestingSyncCommitteeContributionDuty),
-				Duty:                    &testingutils.TestingSyncCommitteeContributionDuty,
-				PostDutyRunnerStateRoot: "29862cc6054edc8547efcb5ae753290971d664b9c39768503b4d66e1b52ecb06",
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/duties/newduty/consensus_not_started.go:30]: remove post root completley?

File: ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go (hunk 3 above)
Comment (alonmuroch) [reply ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go]: same duty should throw and error and shouldn't start IMO

File: ssv/spectest/tests/runner/duties/newduty/finished.go [hunk 15]
```diff
@@ -34,45 +34,42 @@ func Finished() tests.SpecTest {
 		Name: "new duty finished",
 		Tests: []*StartNewRunnerDutySpecTest{
 			{
-				Name:                    "sync committee aggregator",
-				Runner:                  finishRunner(testingutils.SyncCommitteeContributionRunner(ks), &testingutils.TestingSyncCommitteeContributionNexEpochDuty),
-				Duty:                    &testingutils.TestingSyncCommitteeContributionNexEpochDuty,
-				PostDutyRunnerStateRoot: "0ae01aec6a06e42b300b5de874574a19ac2c27a0a32f0ec77f3769a78392065d",
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/duties/newduty/finished.go:40]: removed post root?

File: ssv/spectest/tests/runner/duties/newduty/post_future_decided.go [hunk 16]
```diff
@@ -22,63 +23,60 @@ func PostFutureDecided() tests.SpecTest {
...
 	}
 
 	return &MultiStartNewRunnerDutySpecTest{
 		Name: "new duty post future decided",
 		Tests: []*StartNewRunnerDutySpecTest{
 			{
-				Name:                    "sync committee aggregator",
-				Runner:                  futureDecide(testingutils.SyncCommitteeContributionRunner(ks), &testingutils.TestingSyncCommitteeContributionNexEpochDuty),
-				Duty:                    &testingutils.TestingSyncCommitteeContributionNexEpochDuty,
-				PostDutyRunnerStateRoot: "7f03ca55debd0d503e62817ea81f23f3b01b77b10881c26ba087e5d34b2b5a12",
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/duties/newduty/post_future_decided.go:46]: removed post root?

File: ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post future decided/proposer.json [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+																189,
+																102,
+																128,
+																143
+														],
+														"Graffiti": null
+												},
+												"ID": "AAADAY6ABmVRqBsxglhwntr33R9jzWhqDk24spu7es/mVghnevWlJ9lEjuR4NUheArULwAIAAAA=",
+												"Round": 1,
+												"Height": 50,
```
Comment (alonmuroch) [inline ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post future decided/proposer.json:192]: why the height between the duties are different? here we see height 50, in attester we see 12 and 62?

File: ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post invalid decided/attester.json [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+																189,
+																102,
+																128,
+																143
+														],
+														"Graffiti": null
+												},
+												"ID": "AAADAY6ABmVRqBsxglhwntr33R9jzWhqDk24spu7es/mVghnevWlJ9lEjuR4NUheArULwAAAAAA=",
+												"Round": 1,
+												"Height": 50,
```
Comment (alonmuroch) [inline ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post invalid decided/attester.json:270]: where is this 50 coming from?

File: ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go [hunk 19]
```diff
@@ -15,31 +16,33 @@ func ProposeBlindedBlockDecidedRegular() tests.SpecTest {
...
-					testingutils.TestProposerConsensusDataBytsV(spec.DataVersionBellatrix),
-				), nil),
+					testingutils.TestProposerConsensusDataBytsV(spec.DataVersionCapella),
+				),
+				nil),
 
-			testingutils.SSVMsgProposer(nil, testingutils.PostConsensusProposerMsgV(ks.Shares[1], 1, spec.DataVersionBellatrix)),
-			testingutils.SSVMsgProposer(nil, testingutils.PostConsensusProposerMsgV(ks.Shares[2], 2, spec.DataVersionBellatrix)),
-			testingutils.SSVMsgProposer(nil, testingutils.PostConsensusProposerMsgV(ks.Shares[3], 3, spec.DataVersionBellatrix)),
+			testingutils.SSVMsgProposer(nil, testingutils.PostConsensusProposerMsgV(ks.Shares[1], 1, spec.DataVersionCapella)),
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go]: should we be mixing this in the PR?

File: ssv/spectest/tests/runner/duties/synccommitteeaggregator/proof_all_aggregator.go [hunk 20]
```diff
@@ -22,6 +22,5 @@ func AllAggregatorQuorum() tests.SpecTest {
 			hex.EncodeToString(testingutils.TestingContributionProofsSigned[1][:]): true,
 			hex.EncodeToString(testingutils.TestingContributionProofsSigned[2][:]): true,
 		},
-		PostDutyRunnerStateRoot: "609cb39abfee9a3308c679f5e88de8724215911d9e053c6972e00c32d94e2401",
```
Comment (alonmuroch) [inline ssv/spectest/tests/runner/duties/synccommitteeaggregator/proof_all_aggregator.go:25]: why removed?

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: ssv/spectest/tests/runner/duties/newduty/consensus_not_started.go (hunk 14 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/consensus_not_started.go:30]: Now I want to work to bring them all back since it messes up tests.json

File: ssv/spectest/tests/runner/duties/newduty/finished.go (hunk 15 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/finished.go:40]: ditto

File: ssv/spectest/generate/main.go (hunk 7 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/main.go]: This makes multi-tests output jsons files per subtest.
It makes the world of reviewing states way easier.

It might be o a bit out of scope of this PR.
But it was too hard for me to work w.o this.

File: ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post future decided/proposer.json (hunk 17 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post future decided/proposer.json:192]: Proposers use different slots than other tests. This is because of how our tests mock changes between eth forks.
Currently we produce the proper block version according to the slot.
Other duties don't have entire blocks as CD and they just use constants (12 usually, because this is how it was historically).

As to why you see 12 and 62, I suggest you look into `futureDecide`:
```
		r.GetBaseRunner().State = ssv.NewRunnerState(3, duty)
		r.GetBaseRunner().State.RunningInstance = qbft.NewInstance(
			r.GetBaseRunner().QBFTController.GetConfig(),
			r.GetBaseRunner().Share,
			r.GetBaseRunner().QBFTController.Identifier,
			qbft.Height(duty.Slot))
		r.GetBaseRunner().QBFTController.StoredInstances = append(r.GetBaseRunner().QBFTController.StoredInstances, r.GetBaseRunner().State.RunningInstance)

		futureDecidedInstance := qbft.NewInstance(
			r.GetBaseRunner().QBFTController.GetConfig(),
			r.GetBaseRunner().Share,
			r.GetBaseRunner().QBFTController.Identifier,
			qbft.Height(duty.Slot+50))
		futureDecidedInstance.State.Decided = true
		r.GetBaseRunner().QBFTController.StoredInstances = append(r.GetBaseRunner().QBFTController.StoredInstances, futureDecidedInstance)
		r.GetBaseRunner().QBFTController.Height = qbft.Height(duty.Slot + 50)
		return r
```

Note we create 2 instances. The second instance is at height: `Slot + 50`.

File: ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post invalid decided/attester.json (hunk 18 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*newduty.MultiStartNewRunnerDutySpecTest/new duty post invalid decided/attester.json:270]: https://github.com/bloxapp/ssv-spec/blob/5323e0a6471e58fab7932a3be07bdd6e87e8dd0a/types/testingutils/beacon_node.go#L346

`TestingDutySlot2      = 50`

File: ssv/spectest/generate/state_comparison/*synccommitteeaggregator.SyncCommitteeAggregatorProofSpecTest/sync committee aggregator all are aggregators.json (hunk 8 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*synccommitteeaggregator.SyncCommitteeAggregatorProofSpecTest/sync committee aggregator all are aggregators.json]: The test was broken
When https://github.com/bloxapp/ssv-spec/pull/281 gets merged (soon I hope), you will see a diff.

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json (hunk 9 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: Because I changed the slot that affected the consensus data

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json (hunk 10 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: When we did the Capella support we only changed sanity tests due to time constraints.
So I took the oppertunity to when I went over tests to change Bellatrix to Capella.
This use's Prater's capella slot.

However I am thinking on reverting those changes now to use Bellatrix. This is because:
1. It is not supposed to be in the scope of the PR.
2. Deneb is around the corner and I have a little plan. It maybe actually easier to upgrade to Deneb if we don't change to Capella now (but this is out of scope of the PR)

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json (hunk 11 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: because of slot change

File: ssv/spectest/tests/runner/consensus/future_decided.go (hunk 1 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/future_decided.go]: Yes, it will eventually stop.
But if you know you are behind why wait?

The current behavior makes less sense to me

File: ssv/spectest/tests/runner/consensus/invalid_decided_value.go (hunk 12 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/invalid_decided_value.go:68]: It needs to be correct, because of backwards compatability check you added.
Also note the json take the value from the code...

File: ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go (hunk 13 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go:23]: ditto

File: ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go (hunk 3 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go]: https://github.com/bloxapp/ssv-spec/issues/282

See if you agree with the issue, it is more general than what I wrote in the comment.

File: ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go (hunk 19 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go]: I agree it was a bad call on my side

File: types/testingutils/runner.go (hunk 6 above)
Comment (GalRogozinski) [reply types/testingutils/runner.go:188]: Before we had the same functionused for both `qbft` and `ssv` tests.
What it did was create set of deciding messages from height 0 to `height`.

The issues for using that in `ssv` are:
1. Now that heights are slots it started to create unintentional amount of messages. Changing the tests.
2. The next deciding message stops the instance of the previous set.

File: ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go (hunk 19 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/proposer/propose_blinded_decide_regular.go]: No Capella anymore

File: ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json (hunk 10 above)
Comment (GalRogozinski) [reply ssv/spectest/generate/state_comparison/*tests.MsgProcessingSpecTest/propose blinded decide regular.json]: Capella changes reverted

File: ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go (hunk 2 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go]: opened https://github.com/bloxapp/ssv-spec/issues/282

File: ssv/spectest/tests/runner/duties/newduty/post_wrong_decided.go (hunk 4 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/post_wrong_decided.go:19]: https://github.com/bloxapp/ssv-spec/issues/285

File: ssv/spectest/tests/runner/duties/newduty/post_invalid_decided.go (hunk 5 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/post_invalid_decided.go:37]: https://github.com/bloxapp/ssv-spec/issues/285

File: go.mod [hunk 21]
```diff
@@ -21,6 +21,7 @@ require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/fatih/color v1.13.0 // indirect
 	github.com/gogo/protobuf v1.3.2 // indirect
+	github.com/google/go-cmp v0.5.9 // indirect
```
Comment (alonmuroch) [inline go.mod:24]: why do we need this extra package?

File: ssv/spectest/tests/runner/consensus/future_decided.go (hunk 1 above)
Comment (alonmuroch) [reply ssv/spectest/tests/runner/consensus/future_decided.go]: I'd suggest opening an issue and optimizing it later (as it requires more work/ SIP and tests). 
Unless I'm mistaking I don't think it's super critical

File: ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go (hunk 3 above)
Comment (alonmuroch) [reply ssv/spectest/tests/runner/duties/newduty/duplicate_duty_finished.go]: Shouldn't we solve it in this PR?

File: ssv/spectest/run_test.go [hunk 22]
```diff
@@ -39,6 +40,10 @@ func TestAll(t *testing.T) {
 	wait.Wait()
 }
 
+// TestOne will run a single test.
+func TestOne(t *testing.T) {
+	proposer.ProposeBlindedBlockDecidedRegular().Run(t)
+}
```
Comment (y0sher) [inline ssv/spectest/run_test.go]: By the name and the file location, I was under the impression this function will implement generally running a single test, any single test. though it seems like it runs a specific single test which I'm not sure why and why it is here. is it a mistake? can you explain @GalRogozinski ?

File: ssv/spectest/tests/runner/consensus/future_decided.go (hunk 1 above)
Comment (y0sher) [reply ssv/spectest/tests/runner/consensus/future_decided.go]: I think Alon means it's not urgent, as eventually our desired behavior will happen. adding more functionality like you say might result in bugs. should be thought through before we add something like that.

File: ssv/spectest/tests/runner/consensus/future_decided.go [hunk 23]
```diff
@@ -37,7 +41,7 @@ func FutureDecided() tests.SpecTest {
 						testingutils.TestingCommitMultiSignerMessageWithHeightAndIdentifier(
 							[]*bls.SecretKey{ks.Shares[1], ks.Shares[2], ks.Shares[3]},
 							[]types.OperatorID{1, 2, 3},
-							2,
+							testingutils.TestingDutySlot+1,
```
Comment (y0sher) [inline ssv/spectest/tests/runner/consensus/future_decided.go:45]: why +1?

Comment (y0sher) [reply ssv/spectest/tests/runner/consensus/future_decided.go:45]: ahh because its future.

File: ssv/spectest/run_test.go (hunk 22 above)
Comment (GalRogozinski) [reply ssv/spectest/run_test.go]: this shouldn't be here
I committed it accidentally

Thanks

**Review by y0sher — APPROVED** [review]
at some places you use `testingutils.TestingDutySlot` and some other places you use `qbft.FirstHeight` should this be unified or there's a reason for using one and not the other?

**Comment by GalRogozinski** [comment]
@y0sher 
The only place where I found that I use `qbft.FirstHeight` in this PR is in a QBFT test.
```
@@ -13,7 +13,8 @@ func InvalidValueCheck() tests.SpecTest {
        ks := testingutils.Testing4SharesSet()
        msgs := []*qbft.SignedMessage{
                testingutils.TestingProposalMessageWithIdentifierAndFullData(
-                       ks.Shares[1], types.OperatorID(1), []byte{1, 2, 3, 4}, testingutils.TestingInvalidValueCheck),
+                       ks.Shares[1], types.OperatorID(1), []byte{1, 2, 3, 4}, testingutils.TestingInvalidValueCheck,
+                       qbft.FirstHeight),
        }
```

Technically this PR should only deal with SSV tests. Due to bad SoC the testingutils function that was changed caused a change in QBFT test.

To answer your question, I currently think that the`ssv` section that is in charge of running QBFT instances but don't care about internals should input slots as heights to the instances.  Because that's how the SSV protocol works. For the `qbft` section that defines the instance internal workings it is not prudent.

File: ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go (hunk 13 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/valid_decided_10_operators.go:23]: So currently there are 2 issues:
1. We have that force backwards compatability check
2. Currently we generate `tests.json` before we override the roots... so we must have the correct root in place

File: ssv/spectest/tests/runner/duties/newduty/post_future_decided.go (hunk 16 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/post_future_decided.go:46]: fixed

File: ssv/spectest/tests/runner/duties/synccommitteeaggregator/proof_all_aggregator.go (hunk 20 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/synccommitteeaggregator/proof_all_aggregator.go:25]: added back

File: ssv/spectest/tests/runner/duties/newduty/post_future_decided.go (hunk 16 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/post_future_decided.go:46]: added back

File: ssv/spectest/tests/runner/duties/newduty/finished.go (hunk 15 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/finished.go:40]: I see it is there currently as it should be

File: ssv/spectest/tests/runner/duties/newduty/consensus_not_started.go (hunk 14 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/duties/newduty/consensus_not_started.go:30]: it was added back

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: ssv/spectest/tests/runner/consensus/future_decided.go (hunk 1 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/runner/consensus/future_decided.go]: discussion can continue here:
https://github.com/bloxapp/ssv-spec/issues/280#issuecomment-1628600181

**Review by alonmuroch — APPROVED** [review]
